pkg/image/prune/imagepruner.go

```diff
@@ -488,10 +557,33 @@ func (p *imagePruner) Run(pruneImage ImagePruneFunc, pruneStream ImageStreamPrun
 	allNodes := p.g.NodeList()
 
 	imageNodes := imageNodeSubgraph(allNodes)
+	if len(imageNodes) == 0 {
+		return
+	}
+
+	// we only support a single internal registry, and all images should have the same registry
+	// so we just take the 1st one and use it
+	ref, err := imageapi.ParseDockerImageReference(imageNodes[0].Image.DockerImageReference)
+	if err != nil {
+		glog.Errorf("error determining registry URL from %q: %v", imageNodes[0].Image.DockerImageReference, err)
+		return
+	}
+
+	registryURL := ref.Registry
+	if len(registryURL) == 0 {
+		glog.Errorf("unable to prune - registry URL is missing")
+		return
+	}
+
+	if err := p.registryPinger.ping(registryURL); err != nil {
+		glog.Errorf("error communicating with registry: %v", err)
+		return
+	}
+
 	pruneImages(p.g, imageNodes, pruneImage, pruneStream, pruneManifest)
 
 	layerNodes := layerNodeSubgraph(allNodes)
-	pruneLayers(p.g, layerNodes, pruneLayer, pruneBlob)
+	pruneLayers(p.g, registryURL, layerNodes, pruneLayer, pruneBlob)
```

Comment:
On Jun 30, 2015, at 8:58 PM, Andy Goldstein <notifications@github.com>
wrote:

In pkg/image/prune/imagepruner.go
<https://github.com/openshift/origin/pull/3515#discussion_r33640350>:

>  	pruneImages(p.g, imageNodes, pruneImage, pruneStream, pruneManifest)
>
>  	layerNodes := layerNodeSubgraph(allNodes)
> -	pruneLayers(p.g, layerNodes, pruneLayer, pruneBlob)
> +	pruneLayers(p.g, registryURL, layerNodes, pruneLayer, pruneBlob)

I 100% agree. That's what #3333
<https://github.com/openshift/origin/issues/3333> is about. However, this
PR should eliminate certain problems (it rejects pruning if the registry
isn't reachable, it prunes blobs if there are no streams referencing them)
but it doesn't address the overall issue. I do not believe this PR makes
things worse.

I can modify it so pruneLayers is called first, and if any prune-layer or
prune-blob call fails, don't call pruneImages. That would at least avoid
deleting data from etcd that we'd need to use to determine the layer/blob
deletions to retry.

----

Yes - the concern even then is if the image is repushed it will be in an
invalid state, which means it would be broken to end users.  That's better
than potentially leaving a large chunk of layers in the registry though.

----

Tombstoning would probably need to come as part of the bigger refactor.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3515/files#r33640350>.


----

I think we need to talk through the larger refactor tomorrow.